contracts/ConvictionVoting.sol

```diff
@@ -106,121 +121,150 @@ contract ConvictionVoting is AragonApp, TokenManagerHook {
-        require(!proposal.executed, ERROR_PROPOSAL_ALREADY_EXECUTED);
-        proposal.executed = true;
-        _calculateAndSetConviction(_id, proposal.stakedTokens);
-        require(proposal.convictionLast > calculateThreshold(proposal.requestedAmount), ERROR_INSUFFICIENT_CONVICION_TO_EXECUTE);
-        emit ProposalExecuted(_id, proposal.convictionLast);
+    function executeProposal(uint256 _proposalId) external isInitialized() {
+        Proposal storage proposal = proposals[_proposalId];
+
+        require(_proposalId != ABSTAIN_PROPOSAL_ID, ERROR_CANNOT_EXECUTE_ABSTAIN_PROPOSAL);
+        require(proposal.proposalStatus == ProposalStatus.Active, ERROR_PROPOSAL_NOT_ACTIVE);
+        _calculateAndSetConviction(proposal, proposal.stakedTokens);
+        require(proposal.convictionLast > calculateThreshold(proposal.requestedAmount), ERROR_INSUFFICIENT_CONVICION);
+
+        proposal.proposalStatus = ProposalStatus.Executed;
         vault.transfer(requestToken, proposal.beneficiary, proposal.requestedAmount);
-        if (_withdrawIfPossible && proposal.stakesPerVoter[msg.sender] > 0) {
-            _withdraw(_id, proposal.stakesPerVoter[msg.sender], msg.sender);
-        }
+
+        emit ProposalExecuted(_proposalId, proposal.convictionLast);
+    }
+
+    /**
+     * @notice Cancel proposal #`_proposalId`
+     * @param _proposalId Proposal id
+     */
+    function cancelProposal(uint256 _proposalId) external {
+        Proposal storage proposal = proposals[_proposalId];
+
+        bool senderHasPermission = canPerform(msg.sender, CANCEL_PROPOSAL_ROLE, new uint256[](0));
+        require(proposal.submitter == msg.sender || senderHasPermission, ERROR_SENDER_CANNOT_CANCEL);
+        require(_proposalId != ABSTAIN_PROPOSAL_ID, ERROR_CANNOT_CANCEL_ABSTAIN_PROPOSAL);
+        require(proposal.proposalStatus == ProposalStatus.Active, ERROR_PROPOSAL_NOT_ACTIVE);
+
+        proposal.proposalStatus = ProposalStatus.Cancelled;
     }
 
     /**
-     * @dev Get proposal parameters
-     * @param _id Proposal id
+     * @dev Get proposal details
+     * @param _proposalId Proposal id
      * @return Requested amount
      * @return Beneficiary address
      * @return Current total stake of tokens on this proposal
      * @return Conviction this proposal had last time calculateAndSetConviction was called
      * @return Block when calculateAndSetConviction was called
      * @return True if proposal has already been executed
+     * @return ProposalStatus defining the state of the proposal
      */
-    function getProposal(uint256 _id) public view returns (
-        uint256,
-        address,
-        uint256,
-        uint256,
-        uint64,
-        bool
+    function getProposal(uint256 _proposalId) public view returns (
+        uint256 requestedAmount,
+        address beneficiary,
+        uint256 stakedTokens,
+        uint256 convictionLast,
+        uint64 blockLast,
+        ProposalStatus proposalStatus,
+        address submitter
```

Comment:
Are the namings necessary if we are not assigning them the return values?

Reply:
It doesn't effect performance so I think we can leave them.